ddtrace/internal/processor.py

```diff
@@ -12,18 +15,76 @@
 
 
 @attr.s
-class TraceProcessor(object):
-    _filters = attr.ib(type=List[TraceFilter])
-
-    def process(self, trace):
-        # type: (List[Span]) -> Optional[List[Span]]
-        for filtr in self._filters:
-            try:
-                trace = filtr.process_trace(trace)
-            except Exception:
-                log.error("error applying filter %r to traces", filtr, exc_info=True)
-            else:
-                if trace is None:
-                    log.debug("dropping trace due to filter %r", filtr)
-                    return
+class Processor(six.with_metaclass(abc.ABCMeta)):
+    """A Processor is used to process spans as they are created and finished by a tracer."""
+
+    def __attrs_post_init__(self):
+        # type () -> None
+        """Default post initializer which logs the representation of the
+        Processor at the ``logging.DEBUG`` level.
+
+        The representation can be modified with the ``repr`` argument to the
+        attrs attribute::
+
+            @attr.s
+            class MyProcessor(Processor):
+                field_to_include = attr.ib(repr=True)
+                field_to_exclude = attr.ib(repr=False)
+        """
+        log.debug("initialized processor %r", self)
+
+    @abc.abstractmethod
+    def on_span_start(self, span):
+        # type: (Span) -> None
+        """Called when a span is started.
+
+        This method is useful for making upfront decisions on spans.
+
+        For example, a sampling decision can be made when the span is created
+        based on its resource name.
+        """
+        pass
+
+    @abc.abstractmethod
+    def on_span_finish(self, data):
+        # type: (Any) -> Any
+        """Called with the result of any previous processors or initially with
+        the finishing span when a span finishes.
+
+        It can return any data which will be passed to any processors that are
+        applied afterwards.
+        """
+        pass
+
+
+@attr.s(eq=False)
+class TraceFilterProcessor(Processor):
+    """Processor for applying a ``TraceFilter`` to traces.
+
+    If a filter raises an exception it will be logged at the ``logging.ERROR``
+    level and the trace will be returned.
+    """
+
+    _filter = attr.ib(type=TraceFilter, repr=True)
+
+    def on_span_start(self, span):
+        # type: (Span) -> None
+        return None
+
+    def on_span_finish(self, trace):
+        # type: (Optional[List[Span]]) -> Optional[List[Span]]
+        if not trace:
+            return trace
+
+        trace_id = trace[0].trace_id
+        try:
+            log.debug("applying filter %r to %d", self._filter, trace_id)
+            trace = self._filter.process_trace(trace)
```

Comment:
wait, on second thought why are we doing it this way?

instead of say:

```python
class Processor:
    def on_span_start(self, span): ...
    def on_span_finish(self, span): ...
    def process_trace(self, trace): ...
```

And then we can migrate people from using `process_trace` on their filter class to using `on_span_start`/`on_span_end` instead?

Reply:
Hmm with what we're working toward in #2038 there's no guarantee that a processor will be able to have process_trace called so it'd feel a little wrong to be included in the general API of a processor. Only processors after the `SpansToTraceProcessor` would be able to have process_trace called.